### PR TITLE
sql: remove FK legacy code

### PR DESCRIPTION
**Pull request by RaduBerinde**
#### sql: remove FK fallback

This commit removes the settings that control whether FK operations use the new
optimizer-driven code or if they fall back to the legacy code. After this
change, the legacy code can be removed.

Release note (sql change): Removed the `optimizer_foreign_keys` and
`experimental_optimizer_foreign_key_cascades` session settings (and their
corresponding cluster settings).

#### sql: remove legacy FK code

This change removes all the legacy code around FK checks and cascades, and
anything that becomes unused as a result.

Release note: None

#### lint: ignore unused Cause errors

Errors implement a `Cause()` method that is used by the errors package. The
linter started complaining about unused `Cause` methods; this commit ignores
these errors.

Release note: None

### CODE REVIEW

**Comment by rohany** [comment]
-12,000 😱😱😱

**Comment by RaduBerinde** [comment]
CC @knz for the last commit.

**Comment by RaduBerinde** [comment]
> -12,000 😱😱😱

Heh, about 7000 are just the logictests (that were already copied into the `_opt` variants).

File: pkg/ccl/importccl/load.go [hunk 1]
```diff
@@ -369,12 +369,10 @@ func insertStmtToKVs(
 		if err != nil {
 			return errors.Wrapf(err, "process insert %q", insertRow)
 		}
-		// TODO(bram): Is the checking of FKs here required? If not, turning them
-		// off may provide a speed boost.
 		// TODO(mgartner): Add partial index IDs to ignoreIndexes that we should
 		// not add entries to.
 		var ignoreIndexes util.FastIntSet
-		if err := ri.InsertRow(ctx, b, insertRow, ignoreIndexes, true, row.CheckFKs, false /* traceKV */); err != nil {
```
Comment (RaduBerinde) [inline pkg/ccl/importccl/load.go:377]: Note - this `CheckFKs` wasn't actually doing anything because the inserter was created with `SkipFKs` above.

**Comment by rohany** [comment]
SQL `row/` portions LGTM. So nice to see all of these "delete when optimizer plans cascades" codepaths removed. Are you planning on deleting the DDL portions of old foreign key planning here, or in a future PR (mainly the requiring an index to be created on the origin side, and creating one implicitly).

**Comment by RaduBerinde** [comment]
> Are you planning on deleting the DDL portions of old foreign key planning here, or in a future PR (mainly the requiring an index to be created on the origin side, and creating one implicitly).

In a future PR - once this one is in, I will circle back with the schema team and figure out how to move forward with that.

**Review by knz — COMMENTED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50185)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @lucy-zhang, @pbardea, @RaduBerinde, and @rohany)

___
*[pkg/testutils/lint/lint_test.go, line 1458 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50185#-M9m80J4-uTeIlDn3A4g:-M9m80J4-uTeIlDn3A4h:bdz1k6j) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b8e87de35747e3789b3c974babf7afdc59e78ce/pkg/testutils/lint/lint_test.go#L1458)):*
> ```Go
> 				// Cause is a method used by pkg/cockroachdb/errors (through an unnamed
> 				// interface).
> 				stream.GrepNot(`pkg/.*.go:.* func .*\.Cause is unused`),
> ```

:100: thanks

**Review by thoszhang — COMMENTED** [review]
No objections from me.

> once this one is in, I will circle back with the schema team and figure out how to move forward with that.

Makes sense. I think the code changes required are basically just to put the old index checking/auto-creation code behind a pre-20.2 version gate for backward compatibility, and otherwise remove it. That shouldn't be too complicated. It's mostly a matter of adding tests for both the new permitted FK DDL statements and the actual index-less FK checks themselves.

> Reviewed 32 of 32 files at r1, 33 of 33 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50185#-:-M9saHOE7rgas-P39ZJC:bdqu8yo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @pbardea, @RaduBerinde, and @rohany)

**Review by rytaft — APPROVED** [review]
opt changes :lgtm: 

> Reviewed 26 of 32 files at r1, 1 of 33 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50185#-:-M9tXB0lAydYKlINF_9a:b-yql68r)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @pbardea, @RaduBerinde, and @rohany)

___
*[pkg/sql/logictest/testdata/logic_test/cascade_opt, line 2 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50185#-M9tWkli51ma6uEW8Ywt:-M9tWkli51ma6uEW8Ywu:bs2fpdf) ([raw file](https://github.com/cockroachdb/cockroach/blob/01a780bdc380d428611348a08aff301ae7c1938b/pkg/sql/logictest/testdata/logic_test/cascade_opt#L2)):*
> ```
> # This test times out when run with 3node-tenant.
> # LogicTest: !3node-tenant
> ```

Maybe rename to `cascade`?
___
*[pkg/sql/logictest/testdata/logic_test/fk_opt, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50185#-M9tWrNxE_LCfvVoJXwr:-M9tWrNxE_LCfvVoJXws:bvtu02k) ([raw file](https://github.com/cockroachdb/cockroach/blob/01a780bdc380d428611348a08aff301ae7c1938b/pkg/sql/logictest/testdata/logic_test/fk_opt#L1)):*
> ```
> # LogicTest: local fakedist
> ```

rename to `fk`?
___
*[pkg/sql/opt/exec/execbuilder/testdata/cascade_opt, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50185#-M9tVVVO9R9dXZNzJTnH:-M9tVVVPDYXvNsn_lbmR:b-om0qyl) ([raw file](https://github.com/cockroachdb/cockroach/blob/01a780bdc380d428611348a08aff301ae7c1938b/pkg/sql/opt/exec/execbuilder/testdata/cascade_opt#L1)):*
> ```
> # LogicTest: local
> ```

Maybe you should rename this file to `cascade`?
___
*[pkg/sql/opt/exec/execbuilder/testdata/fk_opt, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50185#-M9tVsHxDNKdJl2asumT:-M9tVsHxDNKdJl2asumU:b-t4dsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/01a780bdc380d428611348a08aff301ae7c1938b/pkg/sql/opt/exec/execbuilder/testdata/fk_opt#L1)):*
> ```
> # LogicTest: local
> ```

and maybe rename this one to `fk`?

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50185)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @knz, @lucy-zhang, @pbardea, @RaduBerinde, @rohany, and @rytaft)

___
*[pkg/sql/opt/exec/execbuilder/testdata/fk_opt, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50185#-M9tVsHxDNKdJl2asumT:-M9uRq3LBoYxp98T2pnK:b8ale6t) ([raw file](https://github.com/cockroachdb/cockroach/blob/01a780bdc380d428611348a08aff301ae7c1938b/pkg/sql/opt/exec/execbuilder/testdata/fk_opt#L1)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

and maybe rename this one to `fk`?
</blockquote></details>

Done. There was a small `fk` file, I consolidated them.
___
*[pkg/sql/logictest/testdata/logic_test/fk_opt, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50185#-M9tWrNxE_LCfvVoJXwr:-M9uRolx5rGobZ011Bsq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/01a780bdc380d428611348a08aff301ae7c1938b/pkg/sql/logictest/testdata/logic_test/fk_opt#L1)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

rename to `fk`?
</blockquote></details>

Done.
___
*[pkg/sql/opt/exec/execbuilder/testdata/cascade_opt, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50185#-M9tVVVO9R9dXZNzJTnH:-M9uRp2EELySoYcLJZg7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/01a780bdc380d428611348a08aff301ae7c1938b/pkg/sql/opt/exec/execbuilder/testdata/cascade_opt#L1)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Maybe you should rename this file to `cascade`?
</blockquote></details>

Done.

**Review by rytaft — APPROVED** [review]
> Reviewed 1 of 32 files at r1, 2 of 33 files at r5, 5 of 8 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50185)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @knz, @lucy-zhang, @pbardea, @RaduBerinde, @rohany, and @rytaft)

**Pull request merged**